contract/src/main.rs

```diff
@@ -420,35 +567,51 @@ pub extern "C" fn approve() {
         OwnershipMode::Transferable => OwnershipMode::Transferable,
     };
 
-    let caller = runtime::get_caller();
-    let token_id = get_named_arg_with_user_errors::<U256>(
+    let holder_mode = get_holder_mode().unwrap_or_revert_with(NFTCoreError::InvalidHolderMode);
+
+    let caller: Key = match holder_mode {
+        NFTHolderMode::Accounts => Key::Account(runtime::get_caller()),
+        NFTHolderMode::Contracts => get_calling_contract_hash().into(),
+    };
```

Comment:
should extract a function `get_verified_holder` and use it here and elsewhere

if the immediate caller is not compatible with the holder_mode, revert.